### PR TITLE
fix data augmentation error. Added new test. 

### LINKED ISSUE #26: loss can't be lower than `1e-10`
Loss can't be lower than `1e-10` at the first iteration in 1 qubit case simulation, even using the same training model in `test_trainerModel`. The `test_trainerModel` can always be passed.

image
image

### PR DESCRIPTION
**Pull request by cfenglv**
Added reinitialization in the training process. Added best model backup during training.

### CODE REVIEW

**Comment by vindolineKis** [comment]
👍👍👍

File: QCircLearning/utils.py [hunk 1]
```diff
@@ -64,3 +64,15 @@ def __call__(self, current_loss):
...
+def reinitialize_network(model):
+    """
+    Re-initializes the parameters of a PyTorch network in place
+    (without constructing a new model).
+
+    model (nn.Module): The model to re-init.
+    """
+    for module in model.modules():
+        if hasattr(module, "reset_parameters"):
+            module.reset_parameters()
```
Comment (Canoming) [inline QCircLearning/utils.py:82]: Why we need to check the attribute `reset_parameters` here? In what case a `module` may not define the `reset_parameters`?

File: QCircLearning/trainerModel.py [hunk 2]
```diff
@@ -130,8 +132,11 @@ def NN_opt(func, x0, callback=None, **kwargs):
                         f"Run ID: {kwargs['run_id']}, Epoch {epoch + 1}/{classical_epochs}, Average Loss: {total_loss:.1e}"
                     )
                     sys.stdout.flush()
+                if early_stopping.counter == 0:
+                    best_model = copy.deepcopy(model)
```
Comment (Canoming) [inline QCircLearning/trainerModel.py]: To clarify the logic here:
```suggestion
                best_model = copy.deepcopy(model) if early_stopping.reset else best_model
```

And define the attribute `reset` in `EarlyStopping`:

```python
@property
def reset(self):
    if self.counter == 0
```

Also, we should move this to the beginning of the loop. Since we will call `model.eval()` regardless, it doesn't make sense to assign `None` to the `best_model`

Comment (Canoming) [reply QCircLearning/trainerModel.py]: Or maybe just keep it at the bottom to confirm the update. But, since we will assign the model at least once, we don't have to define `best_model` outside the loop.

File: QCircLearning/utils.py (hunk 1 above)
Comment (cfenglv) [reply QCircLearning/utils.py:82]: model.modules() will contains activation function or other non-parametric modules that does not have reset_parameters

**Comment by cfenglv** [comment]
Refined code for logic clarification. Optimize best version backup efficiency with only deep copy the state_dict instead of the entire network.

**Review by Canoming — APPROVED** [review]
